apps/internal/oauth/ops/authority/authority.go

```diff
@@ -314,22 +323,56 @@ func (c Client) GetTenantDiscoveryResponse(ctx context.Context, openIDConfigurat
 }
 
 func (c Client) AADInstanceDiscovery(ctx context.Context, authorityInfo Info) (InstanceDiscoveryResponse, error) {
-	qv := url.Values{}
-	qv.Set("api-version", "1.1")
-	qv.Set("authorization_endpoint", fmt.Sprintf(authorizationEndpoint, authorityInfo.Host, authorityInfo.Tenant))
-
-	discoveryHost := defaultHost
-	if TrustedHost(authorityInfo.Host) {
-		discoveryHost = authorityInfo.Host
+	region := ""
+	var err error
+	resp := InstanceDiscoveryResponse{}
+	if authorityInfo.Region != "" && authorityInfo.Region != autoDetectRegion {
+		region = authorityInfo.Region
+	} else if authorityInfo.Region == autoDetectRegion {
+		region = c.detectRegion(ctx)
 	}
+	if region != "" {
+		resp.TenantDiscoveryEndpoint = fmt.Sprintf(TenantDiscoveryEndpointWithRegion, region, authorityInfo.Host, authorityInfo.Tenant)
+		metadata := InstanceDiscoveryMetadata{
+			PreferredNetwork: fmt.Sprintf("%v.%v", region, authorityInfo.Host),
+			PreferredCache:   authorityInfo.Host,
+			Aliases:          []string{fmt.Sprintf("%v.%v", region, authorityInfo.Host), authorityInfo.Host},
+		}
+		resp.Metadata = []InstanceDiscoveryMetadata{metadata}
+	}
+	if region == "" {
+		qv := url.Values{}
+		qv.Set("api-version", "1.1")
+		qv.Set("authorization_endpoint", fmt.Sprintf(authorizationEndpoint, authorityInfo.Host, authorityInfo.Tenant))
+
+		discoveryHost := defaultHost
+		if TrustedHost(authorityInfo.Host) {
+			discoveryHost = authorityInfo.Host
+		}
 
-	endpoint := fmt.Sprintf(instanceDiscoveryEndpoint, discoveryHost)
-
-	resp := InstanceDiscoveryResponse{}
-	err := c.Comm.JSONCall(ctx, endpoint, http.Header{}, qv, nil, &resp)
+		endpoint := fmt.Sprintf(instanceDiscoveryEndpoint, discoveryHost)
+		err = c.Comm.JSONCall(ctx, endpoint, http.Header{}, qv, nil, &resp)
+	}
 	return resp, err
 }
 
+func (c Client) detectRegion(ctx context.Context) string {
+	region := os.Getenv(regionName)
+	if region != "" {
+		region = strings.ReplaceAll(region, " ", "")
+		return strings.ToLower(region)
+	}
+	qv := url.Values{}
+	qv.Set("api-version", defaultAPIVersion)
+	qv.Set("format", "text")
+	resp := ""
+	err := c.Comm.JSONCall(ctx, imdsEndpoint, http.Header{}, qv, nil, &resp)
```

Comment:
design doc indicates that this call should be retried once.

Additionally, do we want to handle discovery here for new versions? Specified as optional in the design doc.

Reply:
https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/358#discussion_r629391772

I was following this discussion and decided to go with the simpler approach of trying it only once. 
I'll update to use the latest API version though as per this doc. https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=linux#supported-api-versions